### DISCUSSION TITLE
RFC: SQL Editor 2.0

### DISCUSSION THREAD

**Discussion by saltcod** [supabase/supabase#14206] [General]
The goal of this RFC (Request for Comment) is to identify how developers are currently using the SQL Editor in the Supabase Studio. In advance of a 2.0 revision, we want to make sure we have a strong sense of how developers are using the tool. 

### Motivation

We’ve been gathering feedback about the SQL Editor from Github Issues, Support tickets and from user interviews. With each new piece of feedback, we learn new ways in which people are using the editor to build their projects.

We also learned that: 

- the way the SQL Editor and SQL Snippets are currently implemented isn’t particularly flexible
- there’s a bunch of obvious pain points that we can easily address in the near term 
- many large projects would benefit from a much more advanced SQL and Snippet interface

Before jumping into solutions, we wanted to gather additional feedback from the community to make sure we were considering a wide array of use cases. 

### What we currently know

- Many developers use the SQL Editor as the primary interface for the Studio. It’s where they write, store and organize all functions and queries related to their project. We’ve seen that many people don’t even use the Table Editor — literally everything is done through the SQL editor
- Many users are heavy users of SQL Snippets. The Snippet interface is pretty inflexible and not ideal for accommodating different workflows at the moment
- There’s some (very) low hanging fruit we can address in the next few weeks—like adding the ability to sort snippets

### What we’d like from you

 We’d love to get your feedback on:

- how you use the SQL Editor in your work?
- if the SQL Editor is core to your work, what does your workflow look like?
- what pain points you run into (both large and small)?
- what you’d change about the editor if you could?
- do you use an external tool (like TablePlus) instead? Why?

**Comment by GaryAustin1** [comment]
A big weakness is the syntax checker and lack of detail on errors.  Many users encounter a syntax error or message that does not show where the error is at.   In helping them I often copy their code to Datagrip and get much more info on where the errors are.

**Reply by typeleven** [reply]
The ability to execute only the currently selected code

A SQL Builder of some sort would be amazing, to get a basic structure to start working with a tweaking would be very useful.

The ability to convert a SQL call to a JS call and vise versa.

**Reply by saltcod** [reply]
Nice one @typeleven heard that last from someone just last week. A Jupyter Notebook style thing.

**Reply by saltcod** [reply]
Quick update @typeleven —
> The ability to execute only the currently selected code

this is now possible! You can run just the selected query via cmd+return or the RUN button!

**Reply by saltcod** [reply]
@typeleven on the other item — what do you mean by SQL Builder? Have you used something like this somewhere else?

**Comment by psteinroe** [comment]
Thanks for looking into possible improvements to the SQL Editor. There are two possible areas of improvement: The editor itself, and the Snippet Management. 

Across both, I would list a few possible use-cases.

## Use Cases
### Writing SQL locally to be later compiled into a migration
This is not very relevant to me, since I am using my local editor to write migrations. Hence, I would not benefit a lot from improvements to lsp features (syntax errors etc) within the editor. 
### Postgres Playground
While writing migrations in my IDE, I sometimes use the sql editor of my local instance to execute quick, playground-like sql statements. Think replit eg develop a custom regexp. 
### Debugging and Exploration
This is done in the production database. There are a lot of cases where I need to find out whether data is corrupted, check counts, find certain rows, etc. Since we develop all our business logic in db functions and triggers, the data is always the first thing to look at when debugging potential issues.
### Performance Analysis
Very close to the previous use case, but comes with a few extra requirements. At the moment, I run a statement like
```sql
begin;
 set local role authenticated;
 set local "request.jwt.claim.sub" to '000006a9-b07e-0000-bd5a-1e420f800000';

 EXPLAIN (ANALYZE, COSTS, VERBOSE, BUFFERS, format json) select id from contact where organisation_id = '82c4d000-28ce-4268-9170-02c29036e000' and (sms is not null or whatsapp is not null or email is not null) order by display_name asc;
rollback;   
```
Hence, I start a transaction block, set the role and the user for that transaction, return the the explain analyse output before rolling back. Finally, I copy the query plan into a visualiser to make my analysis. 

## Feature requests 
### Jupiter Notebook 
All of the above would benefit a lot from a Jupyter Notebook-like experience, where a user can execute single statements. Currently, I find myself commenting all but one statement out, since the result for only one statement is shown. Further, if multiple statements executed at once, each result should be printed. Plus 1 if it's possible to set role + user for the statement. The developer would be able search for the user that should be impersonated if using the authenticated role.

### Snippet Management
A few ideas (unordered):
- snippets in the local cli studio should be persisted (either sync to cloud, or safe to file system somehow)
- creating directories of snippets
- sharing snippets with other developers
- sharing snippets with the community 

That's all I can think of at the moment. Will add more points if something comes up. Thanks for the RFC, looking forward to see what you will come up with!

**Reply by saltcod** [reply]
Thanks for this @psteinroe — I think I reached out on twitter, but in case not:  you can now execute single statements in the editor by selecting it and running cmd+return or hitting RUN.

**Comment by shellscape** [comment]
Here's my suggestion and exchange with the kiwi man https://twitter.com/shellscape/status/1655666295268843533

**Comment by vhpx** [comment]
For @psteinroe's suggestion about snippet management for local development setup, I think it's worth taking a look at the following feature request: https://github.com/orgs/supabase/discussions/13331

**Reply by psteinroe** [reply]
That’s such a great idea! Thanks for sharing. ♥️

**Reply by saltcod** [reply]
Yep nice one, thanks!

**Comment by hidech** [comment]
I'm loving to use Supabase SQL editor.
Please consider to add ability of:
- Auto generate title of SQL from the first line of code.
- Organize with folders
- Arbitrary sort of SQLs

I'm also expecting to add some functionality such as coloring, code completion, especially error correction and complement existing tables or column names. Improvement of AI integration is also good.
Happy coding!

**Reply by hidech** [reply]
BTW, I wish I could write Supabase edge functions also in the Supabase studio!

**Reply by Mikestart** [reply]
Hey, I've seen your post on flutterrflow community regarding the forgot password. Did you manage to find a solution ? 
Thanks for your help.

**Comment by Gabrielpanga** [comment]
One cool thing would be to run the query only on the selected text query, if you have multiple queries you need to remove/comment on the ones you don't want to use.

<img width="269" alt="image" src="https://github.com/supabase/supabase/assets/1649924/bc73396a-4f31-41dd-a380-b1bfcbcc21c5">

Also exporting rows to JSON/INSERTS could help.

**Reply by saltcod** [reply]
This should work @Gabrielpanga . You should be able to select a query and run it.

**Reply by Gabrielpanga** [reply]
Oh cool, I was doing it with a click with run rather than (cmd + enter). 

<img width="401" alt="image" src="https://github.com/supabase/supabase/assets/1649924/d7c442cf-c0b1-45e9-b78a-543c8883de07">

**Reply by saltcod** [reply]
Now works by just clicking RUN!

**Comment by dthyresson** [comment]
Currently, the SQL Editor runs as "postgres" --  a high level of access. 

A feature that let you execute queries as 

* other roles
* roles that have specific RLS policies granted to them
* pass JWT token to impersonate a user (to test RLS)

could help test that the queries behave as expected in those either app auth and database auth contexts.

Such a feature is not unlike the what GraphiQL lets you pass the auth header to execute those queries on as an anonymous user or authenticated user.

Eventually, if the Studio is used in a team context, one might want certain dashboard/studio users to only have SQL Editor access under certain permissions. This feature could possible help there as well.

**Comment by GaryAustin1** [comment]
In another discussion:  [Folder Structure for SQL Snippets](https://github.com/orgs/supabase/discussions/14342)

**Comment by saltcod** [comment]
Make copying multiple values easier: 
https://twitter.com/KennethCassel/status/1655623816461135895

**Comment by saltcod** [comment]
Cancel currently running queries
https://twitter.com/steveruizok/status/1655674188756033540

**Comment by saltcod** [comment]
Add explain analyze history for perf benchmarking: 
https://twitter.com/hannyOfsaris/status/1655692537015971840

**Comment by saltcod** [comment]
Copy a row as a SQL insert statement
https://twitter.com/austin_malerba/status/1655627098290106382

**Comment by saltcod** [comment]
SQL Snippet history / roll back:
https://twitter.com/ntkris/status/1655626560920059904

**Comment by saltcod** [comment]
Add a linter to Monaco: 
https://twitter.com/rbkayz/status/1655628298473570305

**Comment by saltcod** [comment]
Tag scripts that are RPC fns: 
https://twitter.com/rbkayz/status/1655628298473570305

**Comment by saltcod** [comment]
Link to Database → functions so they can be edited in the SQL Editor
https://twitter.com/indie_daniel/status/1655632302360125453

**Reply by JulienLecoq** [reply]
I feel the same big pains with this:

> 2. Maybe (and this might be a long shot) there is a way to simplify/automate the process of updating function signatures. I use PostGIS functions and therefore am quite dependent from db functions in general, but I don‘t want to create „_v2“ functions all the time when I add a new argument just to be able to delete the old function an rename it back. Maybe I‘m doing something wrong though.

Also, another thing that I find very painful, is the lack of global search across all files of the SQL Editor. It's very hard to see/remember where each function is used.

**Comment by anngbaum** [comment]
we have certain SQL queries that make many, many http requests using the pg_net extension (doing things like scraping thumbnails, or sending push notifications, as examples). those queries frequently time out if we make too many requests in one go. 

**possible improvements:**
1. ability to temporarily adjust the query timeout from the UI! or somehow distinguish SQL editor-generated requests from client-generated requests (which we do want to keep time limited more conservatively).
2. showing any pg_log statements generated by that query in the interface would be a huge help as well. this would allow us to easily see how far in the query we get before we time out, and fine tune based on that.

**Reply by garylake** [reply]
@anngbaum hey, I had a similar issue where I'm calling my own edge function from pgsql-http and in here I'm calling other end points/services – I'm having to rate limit in the edge function so as not to upset the other services' rate limits that I'm calling, so I was getting timeout issues (5001ms seems to be the default). 

I don't know if this will work for pg_net but assuming it's also curl under the hood this solved the issue for me:

Add to top of you plpgsql function right after 'begin':
```
begin 
  perform http_set_curlopt('CURLOPT_TIMEOUT', '60'); -- this is in actual seconds, not milliseconds
  perform http_set_curlopt('CURLOPT_TCP_KEEPALIVE', '60');
```

**Reply by garylake** [reply]
Hmmm, looking at this again the solution I've given is specific to pgsql-http

I found this in the pg_net docs:
-- WARNING: this is currently ignored, **so there is no timeout** <<<< no timeout supposedly...
    -- the maximum number of milliseconds the request may take before being cancelled
    timeout_milliseconds int default 1000

But it sounds like you are getting a timeout anyway... 

Maybe try pgsql-http instead and then you can set your own timeout config?

**Reply by anngbaum** [reply]
thanks @garylake! will check that out – though I think our issue might actually be a little different. our individual http requests aren't timing out, but rather the query when executed over 100s of rows (or ideally more) times out due to the query timeout mentioned [here](https://supabase.com/docs/guides/database/timeouts).

I can clearly alter the timeout length as mentioned there, but do feel nervous with that solution as I wouldn't want to accidentally initiate some long-running query that takes over. there may be more granularity than I'm aware of in terms of `roles` but generally would love just an easy **temporary** way to extend for specific known queries!

**Reply by anngbaum** [reply]
as I'm writing this I do realize that I could just include a `set statement_timeout` line above and below my queries, so this request is probably more of a UI wrapper suggestion!

regardless, would still love some sort of easily viewable query log statements to just see how far the query gets/for debugging purposes.

**Reply by garylake** [reply]
Nice one, hope you get it sorted! 

You may have helped me actually, elsewhere I have an rpc call that the user can initiate which can take longer than the default timeout – it's not a problem right now but it's a bulk-import type function which could take longer than the default in the future, so thanks for pointing me to this ☺️

**Reply by garylake** [reply]
Wouldn't mind picking your brains some time on thumbnail scraping, and I assume you're using Supabase storage? It's something I need to tackle soon and I'm in two minds about using edge functions or not, as I previously hit a dead end when trying to do it in SQL using pgsql-http...

**Reply by anngbaum** [reply]
yup – happy to discuss, we've spent some time fine tuning that. for our flow we attempt to scrape as part of an `insert` trigger on a table in question, but have to be pretty conservative because if the transaction times out, we don't want to rollback the insert action. so we assume some number will fail, then run a pg_cron job every 10 minutes to retry any failed. and yes we're using Supabase storage, though actually built this before edge functions so implemented with a Google Cloud Function though I imagine it's similar.

**Comment by derek-rein** [comment]
I would like to see a type-safe query builder alongside the Monaco sql code editor. Something like Kysely, or even a graphical query builder using something like litegraph.js. It would be great to be able compose reusable SQL; in my case I have a number of RPC functions that use the same arguments and keeping them in sync is a pain point. It would also be nice to see dependent views/functions and be able to automatically drop them when making changes to the underlying.

**Reply by saltcod** [reply]
Thanks @derek-rein! Great suggestions. Does this help in the meantime? https://supabase-schema.vercel.app/

**Reply by derek-rein** [reply]
@saltcod Thats the right idea, I envisaged something similar to that where you would compose queries visually with a node per SQL keyword. However after considering this a bit more, my sense is that a typescript based query builder would be the way to go for faster, composable, type-safe query building. I have many complex views and functions in my project, beyond the limits of a good DX with raw SQL.

**Comment by garylake** [comment]
## Small bits or +1s
* 'Order by' dropdown for the side column – alphabetical is my preferred/default but it would be nice to order by most recent and created at, for example...
* Resizeable side column. although I figure this is an app-wide request
* Folders for queries in side col
* Query history/versioning 
* Cancel queries (yes I've accidentally had an infinite loop or two)

## Live editing/reload
Google docs style live editing/real time saving. It would be nice to be able to have two queries open as I sometimes want to reference something I've done elsewhere. But I've been caught out before though where I accidentally have the same query open in multiple windows and I've accidentally overwritten one with an older version.

Having this work like a google doc where it doesn't matter where I'm editing, it syncs the versions back into the editor so if I edit in one location it updates in the other to eliminate this risk. 

Which also leads me to...

## Editor tabbed interface
I'm constantly switching between queries so a tabbed interface for the editor would be amazing (although switch tab short cuts would need to switch editor tabs not browser tabs if editor has focus, could get messy I guess)

## Permanent scratch pad/query playground
I personally have a query called '~SQL Scratch Pad' that always sits at the top of my list of queries and I just use it like an SQL terminal when I need to noodle or run quick commands. This is to save me having loads of 'Untitled query' items in my list.  

## Context aware AI
I don't really know if the new 'AI create query' can see/understand my project (the one time I used it, I felt not). This would be a game changer if it could see your whole project/schema etc. I use ChatGPT a lot but I find it exhausting having to work through giving it the context of quite a complex schema and it would be great if it could actually see my project and I can focus on the actual problem I'm facing, not trying to explain my schema. I don't use the AI you've built into Supabase because I already pay GPT Plus and usually need GPT4 for the level of questions I'm asking. 

Even if I have to provide/use my own GPT API access or something like that, would be fine. 

GPT being able to help with RLS policies, esp when there's chunky joins/relationships involved would be amazing! 

## Access to storage from plpgsql
I can probably achieve most of what I need using edge functions and pgsql-http, BUT there are instances when being able to access storage from sql would be handy. 

i.e. my users can upload avatars but when they close their account, I need to use an edge function called via pgsql-http just to delete any media associated with their account. 

select * from storage.delete(filePath, bucket) or something like that would be amazing.

**Reply by garylake** [reply]
Another small bit: when clicking collapse results, still show a single line's worth rather than none as when collapsing, I still want to so see if there's an error or success...

**Reply by saltcod** [reply]
Also a good one!

**Reply by garylake** [reply]
Small bit: Being able to zoom SQL editor text size independently of zooming in browser. I.e. I use Cmd+/Cmd- a lot to zoom as the default editor font is too small for my screen bit I don't actually want to zoom the whole studio UI.

Small dropdown in the editor that lets specify an editor font size would be nice, with a keyboard shortcut too?

**Comment by davetron5000** [comment]
I have meager needs:

* **Ability to turn off the autocomplete** and syntax highlighting. It's almost always wrong and gets in the way of me trying to type.  I'm not against improving it for others, but a text field would be better than what is there now.
* **Ability to run as different roles.**  Right now, the only way to make a view respect RLS is to set the owner to `authenticated` which has the side effect of not returning any rows when run from the editor.
* **Easier ability to see the complete and full results of a query.**  ~Right now, there is truncation of any value over a few characters, no tooltip that shows the full value and no way to resize columns or change the view.  Even allowing the results to wrap in order to fit a width would be better than what there is now~ (you can resize the columns) Something like `\x auto` in `psql` would be nice, but resizing works to get access to the data.

**Reply by saltcod** [reply]
Thanks @davetron5000!

On the last one. Are you not able to resize columns in the results table? 
https://share.cleanshot.com/tBt1yRLW

**Reply by davetron5000** [reply]
Hmm, that does seem to work now.  I swear it has never worked, but it's possible since there is no indication that it's possible (e.g. different cursor) I didn't do it properly.

**Reply by dacameron-proactive** [reply]
> Ability to turn off the autocomplete

Just this. Poss make it triggerable (CTRL-SPACE?) instead of automatically completing the wrong thing as I type. But just being able to switch it off entirely would be a huge improvement. It's a real PitA.

**Comment by peachp** [comment]
No specific comment yet (though I surely had some while using it), but I also use it a lot and generally I'm happy to see that even though Supabase is about  wrapping Postgres into a nice GUI, you guys are still improving such "low level" capabilities.

Oh I have one:
- **Ability to save scripts on account level**, rather than just project level: I recently deleted a project with loads of useful scripts. Another argument is that you often need to use the same scripts across different projects, esp. if you have dev / staging / prod like we do, all hosted on SB, but even if it's totally different projects with different schemas, there is a lot of system scripts (e.g. listing FKs) that I use across projects.

**Reply by saltcod** [reply]
Yup! Def our most-requested feature. Thanks!

**Comment by bombillazo** [comment]
Hey, it would be great to share scripts across users in the same account.

**Reply by mztmoraes** [reply]
that's a good idea!

**Reply by saltcod** [reply]
Hey all! Shipped this exact feature a couple of weeks back. Click the contextual menu on a snippet and you'll be able to share it

**Comment by typeleven** [comment]
When making a SQL statement once you run the statement if the statement still is untitled have ai generate a name for it much like how ChatGPT names your conversations.

**Reply by saltcod** [reply]
That's live now @typeleven! Optional though, via the rename snippet modal.
